core/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java

```diff
@@ -230,8 +229,9 @@ private void processIndexShardSnapshots(ClusterChangedEvent event) {
                 // running shards is missed, then the snapshot is removed is a subsequent cluster
                 // state update, which is being processed here
                 for (IndexShardSnapshotStatus snapshotStatus : entry.getValue().shards.values()) {
-                    if (snapshotStatus.stage() == Stage.INIT || snapshotStatus.stage() == Stage.STARTED) {
-                        snapshotStatus.abort();
+                    final IndexShardSnapshotStatus.Stage stage = snapshotStatus.asCopy().getStage();
+                    if (stage == Stage.INIT || stage == Stage.STARTED) {
+                        snapshotStatus.moveToAborted("snapshot has been removed in cluster state, aborting");
```

Comment:
Would it make sense to change `moveToAborted` so that it only does this if the shard is in INIT or STARTED (rename it to `abortIfNotCompleted`?).

Reply:
That's an interesting suggestion, thanks